agent/xds/listeners.go

```diff
@@ -790,56 +746,49 @@ func (s *Server) makeMeshGatewayListener(name, addr string, port int, cfgSnap *p
 
 func (s *Server) makeUpstreamListenerForDiscoveryChain(
 	u *structs.Upstream,
+	address string,
 	chain *structs.CompiledDiscoveryChain,
 	cfgSnap *proxycfg.ConfigSnapshot,
 	tlsContext *envoyauth.DownstreamTlsContext,
 ) (proto.Message, error) {
-	cfg, err := ParseUpstreamConfigNoDefaults(u.Config)
-	if err != nil {
-		// Don't hard fail on a config typo, just warn. The parse func returns
-		// default config if there is an error so it's safe to continue.
-		s.Logger.Warn("failed to parse", "upstream", u.Identifier(), "error", err)
-	}
-	if cfg.ListenerJSON != "" {
-		s.Logger.Warn("ignoring escape hatch setting because already configured for",
-			"discovery chain", chain.ServiceName, "upstream", u.Identifier(), "config", "envoy_listener_json")
-	}
-
-	addr := u.LocalBindAddress
-	if addr == "" {
-		addr = "127.0.0.1"
+	if address == "" {
+		address = "127.0.0.1"
 	}
-
 	upstreamID := u.Identifier()
+	l := makeListener(upstreamID, address, u.LocalBindPort)
 
-	l := makeListener(upstreamID, addr, u.LocalBindPort)
-
-	proto := cfg.Protocol
-	if proto == "" {
-		proto = chain.Protocol
-	}
-
-	if proto == "" {
-		proto = "tcp"
+	cfg := getAndModifyUpstreamConfigForListener(s.Logger, u, chain)
+	if cfg.ListenerJSON != "" {
+		return makeListenerFromUserConfig(cfg.ListenerJSON)
 	}
 
 	useRDS := true
 	clusterName := ""
-	if proto == "tcp" {
+	if chain == nil || chain.IsDefault() {
+		dc := u.Datacenter
+		if dc == "" {
+			dc = cfgSnap.Datacenter
+		}
+		sni := connect.UpstreamSNI(u, "", dc, cfgSnap.Roots.TrustDomain)
+
+		useRDS = false
+		clusterName = CustomizeClusterName(sni, chain)
+	} else if cfg.Protocol == "tcp" {
```

Comment:
```suggestion
		clusterName = CustomizeClusterName(sni, chain)
		
	} else if cfg.Protocol == "tcp" {
```

This block is a bit crowded, suggesting adding a newline.